src/metric.jl

```diff
@@ -32,54 +28,39 @@ struct DiagEuclideanMetric{T<:Real,A<:AbstractVector{T}} <: AbstractMetric{T}
     _temp   ::  A
 end
 
-Base.show(io::IO, dem::DiagEuclideanMetric) = print(io, _string_diag(dem.M⁻¹))
+function DiagEuclideanMetric(M⁻¹::AbstractVector{T}) where {T<:Real}
+    return DiagEuclideanMetric(M⁻¹, sqrt.(M⁻¹), Vector{T}(undef, size(M⁻¹, 1)))
+end
 
 # Create a `DiagEuclideanMetric` with a new `M⁻¹`
-function (dem::DiagEuclideanMetric)(M⁻¹::A) where {T<:Real,A<:AbstractVector{T}}
-    return DiagEuclideanMetric(dem.dim, M⁻¹)
-end
+(dem::DiagEuclideanMetric)(M⁻¹::AbstractVector{<:Real}) = DiagEuclideanMetric(M⁻¹)
 
-function DiagEuclideanMetric(θ::A) where {T<:Real,A<:AbstractVector{T}}
-    dim = length(θ)
-    return DiagEuclideanMetric(dim, ones(T, dim))
-end
+Base.show(io::IO, dem::DiagEuclideanMetric) = print(io, _string_diag(dem.M⁻¹))
 
-function DiagEuclideanMetric(θ::A, M⁻¹::A) where {T<:Real,A<:AbstractVector{T}}
-    return DiagEuclideanMetric(length(θ), M⁻¹)
+function Base.getproperty(dem::DiagEuclideanMetric, d::Symbol)
+    return d === :dim ? size(getfield(dem, :M⁻¹), 1) : getfield(dem, d)
 end
 
-function DiagEuclideanMetric(dim::Int, M⁻¹::V) where {T<:Real,V<:AbstractVector{T}}
-    @assert dim == length(M⁻¹)
-    return DiagEuclideanMetric(dim, M⁻¹, sqrt.(M⁻¹), zeros(T, dim))
-end
 
-struct DenseEuclideanMetric{T<:Real,AV<:AbstractVector{T},AM<:AbstractMatrix{T}} <: AbstractMetric{T}
-    dim     :: Int
+struct DenseEuclideanMetric{AV<:AbstractVector, AM<:AbstractMatrix} <: AbstractMetric
     # Inverse of the mass matrix
     M⁻¹     ::  AM
     # U of the Cholesky decomposition of the mass matrix
-    cholM⁻¹ ::  UpperTriangular{T,AM}
+    cholM⁻¹ ::  UpperTriangular
     # Pre-allocation for intermediate variables
     _temp   ::  AV
 end
 
-Base.show(io::IO, dem::DenseEuclideanMetric) = print(io, _string_diag(diag(dem.M⁻¹)))
-
-# Create a `DenseEuclideanMetric` with a new `M⁻¹`
-function (dem::DenseEuclideanMetric)(M⁻¹::A) where {T<:Real,A<:AbstractMatrix{T}}
-    return DenseEuclideanMetric(dem.dim, M⁻¹)
+function DenseEuclideanMetric(M⁻¹::AbstractMatrix{T}) where {T<:Real}
+    _temp = Vector{T}(undef, size(M⁻¹, 1))
+    return DenseEuclideanMetric(M⁻¹, cholesky(Symmetric(M⁻¹)).U, _temp)
 end
 
-function DenseEuclideanMetric(θ::A) where {T<:Real,A<:AbstractVector{T}}
-    dim = length(θ)
-    return DenseEuclideanMetric(dim, Matrix{T}(I, dim, dim))
```

Comment:
I suggest we provide the interface `AbstractMetric(d::Int)` to allow users initialise with identity `M⁻¹` for both `DenseEuclideanMetric` and `DiagEuclideanMetric`

Reply:
Good shout, will implement.

Reply:
Happy to merge it when this is done.